### PR TITLE
chore(train): 色々整備

### PR DESCRIPTION
**Pull request by sevenc-nanashi**
## 内容

- 学習に必要なものを除いて削除します。
- 学習用のツールを追加します。

## 関連 Issue

- ref: #5 
- closes: #6

## スクリーンショット・動画など

（なし）

## その他

（なし）

### CODE REVIEW

**Review by Hiroshiba — COMMENTED** [review]
ちょっといくつか転ばぬ先の杖があるのでリストアップしてみます！！

機械学習は常に結果が不確定で、かつ「もう一度やる」のに時間がかかることが多いです。
なので間違えないようにいろいろ工夫すると良いと思います！

* モデルだけの出力ではなく、pth・lossのログ（tfevents）・設定パラメータを保存する
	* ディレクトリを指定できる形にして、その下に全部保存すると便利
	* 人によってはその時の学習コードを全部コピーする人もいる
	* とりあえずargparseの値だけでも保存しておくと良い
	* どうせ勘違いで間違えるので･･･（N敗）
	* OSSの場合、他の人が改善するときの参考になる
* 設定パラメータはconfig.yamlで指定する形にすると便利
	* argparseにすると無限に増えていく
	* そのconfig.yamlを保存するようにすると良い
	* 個人的にはいつも`python train.py [config.yamlのパス] [outputディレクトリのパス]`で統一してます
	* configはそのままdataclassにすると便利
	* あとバリデーションもかけておくと良い、タイポに気づける
	* 最終的に[こういう感じ](https://github.com/Hiroshiba/yukarin_sosfd/blob/main/yukarin_sosfd/config.py)になる
* commit idが勝手に書かれるコードにしておくと便利
	* 間違って別のブランチで起動してたりしても気付ける
	* あとでコードにバグに気づいたとき、このモデルがバグが入る前のコードなのかがわかる
	* [この辺り](https://github.com/Hiroshiba/yukarin_sosfd/blob/6a29a9c35d8aa25b5e3092087963d9c29770b57c/train.py#L28-L33)の感じとか
* deviceの指定をargparseでやるかconfig.yamlでやるかは好み
	* あとあとAWSとかで起動したくなってきたとき、全部がconfig.yamlにあったほうが楽ではある
	* 僕は学習タスクを回すランチャーを別で作ってて、そこでconfig.yamlを構築してtrain.pyに渡してます
	* k8sとかAWSとかGCPとかで取り回しがしやすくなる。
* ３年後でもみんな使ってるライブラリで構築したほうが良い
	* ３年後に改善だのが必要になったとき、戻ってこないといけない
	* 変な環境で作ってると再現がとっっってもめんどくさい
	* たぶんtaskfileは辞めたほうが良い、他はメリットも大きいので大丈夫
* outputディレクトリ以外全部configに入れておくといろいろ便利
	* シード値・trainとevalの比率などなど
* 将来の拡張
	* 今回は簡単なタスクだから無いけど、将来的に必要になってくる要件はこんな感じ
	* 途中から開始できるようにする
		* 学習に１週間かかるとか余裕ででてくる
		* 途中で止めて再開できるようにする
		* optimizerなどを適切に保存することになる、結構難しい
	* パラメータ違いを複数回す
		* 例えばDIMやバッチサイズなど、わりと品質に寄与する
		* いろいろなパターンを一気に回すこともある
		* ディレクトリ名の名付けがめんどくさくなる
		* なので学習ランチャー側で名前を自動決定する機構を作ることになる
		* 僕はf-stringとパラメータから自動的に名前が決まるように作ってます
	* 学習途中に評価指標を計算してtensorboardにプロットする
		* 最終的に評価するのではなく、評価も学習中にやっちゃう感じ
		* train.pyがeval.pyを叩くのではなく、Evaluatorを作って両方がそれを参照すると良い
		* 僕は[こんな感じ](https://github.com/Hiroshiba/yukarin_sosfd/blob/main/yukarin_sosfd/evaluator.py)にしてます
	* pip installできるように作っておく
		* これは好みだけど、このモデルを他でサクッと使えるようにしたくなる時がある
		* pip installしたら入るようにしておくと後々楽
		* [こういうsetup.py](https://github.com/Hiroshiba/yukarin_sosfd/blob/main/setup.py)を書いとくだけで入る（今回は不要そう）

とりあえずこんな感じですかね！！！
気に入ったところとかあれば参考にしていただければ 🙏

File: train/README.md [hunk 1]
```diff
@@ -21,6 +21,12 @@ task train-cuda -- --label [label]
 task tensorboard
 ```
 
+### データセットの結合
+
+```bash
+task merge -- ./vendor/dataset_01.jsonl ./vendor/dataset_02.jsonl --output ./vendor/dataset_merged.jsonl
+```
```
Comment (Hiroshiba) [inline train/README.md:28]: これはdataset側にあったほうが良いかも･･･？

**Comment by sevenc-nanashi** [comment]
設定とかをconfig.ymlで管理するようにシマした。

File: train/src/train.py [hunk 2]
```diff
@@ -223,21 +214,41 @@ def train():
...
-        train_ds,
-        batch_size=batch_size,
-        shuffle=True,
-        collate_fn=partial(collate_fn, device=device),
+    output_dir = os.path.join(
+        "outputs", datetime.now().strftime(f"%Y_%m_%d_%H_%M_%S_{config_name}")
+    )
+
+    print(f"Output dir: {output_dir}")
+    os.makedirs(output_dir, exist_ok=True)
```
Comment (Hiroshiba) [inline train/src/train.py:232]: （このPRじゃなくて良いけど）
pathlibにしたいかも。

**Review by Hiroshiba — COMMENTED** [review]
ちょっと１回TODOコメントいっぱい書くとかしてマージしたいかもです！

**Comment by sevenc-nanashi** [comment]
とりあえずこんな感じ...？

**Review by Hiroshiba — COMMENTED** [review]
良い感じ！！
もうほとんどLGTMです！！

File: train/src/evaluator.py [hunk 3]
```diff
@@ -1,24 +1,32 @@
+from typing import TYPE_CHECKING
 import torch
 from torcheval.metrics import BLEUScore
 from tqdm.auto import tqdm
 
+if TYPE_CHECKING:
+    from train import Model, MyDataset
```
[the remark this answers is not in the document]
Comment (sevenc-nanashi) [reply train/src/evaluator.py:7]: > TYPE_CHECKING要らない気も･･･？

循環importですね...

[the remark this answers is not in the document]
Comment (Hiroshiba) [reply train/src/evaluator.py:7]: あ～～～なるほどです！！
これだと理由がわからないので、問題が解消しても消して良いのか不明かもですね。
とりあえずFIXMEコメント書いたりして意図をわかるようにするのはどうでしょうか？

File: train/src/evaluator.py [hunk 4]
```diff
@@ -1,24 +1,32 @@
...
         with torch.no_grad():
             for eng, kata in tqdm(self.dataset, desc="Evaluating"):
                 res = model.inference(eng)
                 pred_kana = tensor2str(res)
-                kana = [[tensor2str(k) for k in kata]]
+                kana = (
+                    [[tensor2str(kata)]]
+                    if isinstance(kata, torch.Tensor)
+                    else [[tensor2str(k) for k in kata]]
+                )
```
[the remark this answers is not in the document]
Comment (sevenc-nanashi) [reply train/src/evaluator.py:29]: そもそもこれをやってる理由がe2kの元の学習データが`{word: string, kata: string[]}`だからなので、`{word: string, kata: string}`にしちゃえば色々とキレイになるかも...？

とりあえずこのPRでは一旦これで行こうと思います。

File: train/src/evaluator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sevenc-nanashi) [reply train/src/evaluator.py]: FIXMEを足しました。

File: train/src/evaluator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hiroshiba) [reply train/src/evaluator.py:31]: > そもそもこれをやってる理由がe2kの元の学習データが{word: string, kata: string[]}だからなので、{word: string, kata: string}にしちゃえば色々とキレイになるかも...？

これちょっと元の実装の気持ちを知らないのですが、多分複数の読みがありえる時に対応してるんですよね。
そこどうするかなと思います！

[the remark this answers is not in the document]
Comment (sevenc-nanashi) [reply train/src/evaluator.py:31]: トークン数とかの事情も考えると複数読みも出すのは厳しいと思うので複数読みは想定しない方向でやっても良いような気がします。

File: train/src/train.py [hunk 5]
```diff
@@ -87,7 +89,7 @@ def inference(self, src):
 
 
 class MyDataset(Dataset):
-    def __init__(self, path, device):
+    def __init__(self, path, device, max_words: int | None = None):
```
[the remark this answers is not in the document]
Comment (sevenc-nanashi) [reply train/src/train.py]: デフォルト値を消しました。

File: train/src/evaluator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hiroshiba) [reply train/src/evaluator.py:31]: 英→カタカナは本質的に１：Ｎ対応だと思うので、それを反映したデータ構造にするのは割とアリな気もします。
とはいえ学習データも作る今の方針だと恩恵がないのはそうなので、まあどっちもアリ！

とりあえず方針はどちらでもOKとして、TODOコメント対応をお願いできると。。

**Review by Hiroshiba — APPROVED** [review]
LGTM！！！

直感２年分くらいの試行錯誤をショートカットできた気がします！！

File: train/src/train.py [hunk 6]
```diff
@@ -223,21 +214,41 @@ def train():
...
     print(f"Batch size: {batch_size}")
 
-    train_dl = DataLoader(
-        train_ds,
-        batch_size=batch_size,
-        shuffle=True,
-        collate_fn=partial(collate_fn, device=device),
+    output_dir = os.path.join(
+        "outputs", datetime.now().strftime(f"%Y_%m_%d_%H_%M_%S_{config_name}")
+    )
```
[the remark this answers is not in the document]
Comment (Hiroshiba) [reply train/src/train.py:222]: これ、outputディレクトリを指定する形にして、無指定だったらこの値が入る、みたいなのが落とし所かも。

[the remark this answers is not in the document]
Comment (sevenc-nanashi) [reply train/src/train.py]: 引数で上書きできるようにしました。

**Comment by sevenc-nanashi** [comment]
マージします。
